Adding API Specs for Long Term Retention V2

trgrie opened the pull request:
This change adds the swagger documentation for the Long Term Retention V2 feature along with examples for each of the APIs.. This includes the following APIs:

Get an LTR Backup
Get all LTR Backups under a database
Get all LTR Backups under a server
Get all LTR Backups under a location
Delete an LTR Backup
Set a database's LTR Policy
Get a database's LTR Policy

This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process. 

### PR information
- [X] The title of the PR is clear and informative.
- [X] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [X] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [X] If applicable, the PR references the bug/issue that it fixes.
- [X] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [X] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [X] My spec meets the review criteria:
  - [X] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [X] The spec follows the guidelines described in the [Swagger checklist](../documentation/swagger-checklist.md) document.
  - [ ] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

trgrie committed:
Adding API Specs for Long Term Retention V2

This change adds the swagger documentation for the Long Term Retention V2 feature along with examples for each of the APIs.. This includes the following APIs:

Get an LTR Backup
Get all LTR Backups under a database
Get all LTR Backups under a server
Get all LTR Backups under a location
Delete an LTR Backup
Set a database's LTR Policy
Get a database's LTR Policy

anuchandy commented:
@trgrie please update the readme file with an entry for this spec so that CI can validate it.

anuchandy commented on specification/sql/resource-manager/Microsoft.Sql/preview/2017-03-01-preview/longTermRetention.json:
```diff
@@ -0,0 +1,724 @@
+        "yearlyRetention": {
+          "description": "The yearly retention policy for an LTR backup.",
+          "type": "string"
+        },
```
Are these fields `weeklyRetention, monthlyRetention, yearlyRetention `  always going to be integer? asking because in example those are numbers inside double quotes

trgrie replied:
These will be strings in an ISO 8601 standard meaning P1Y is a period of 1 year. I'm updating the description to include this information.

anuchandy commented on specification/sql/resource-manager/Microsoft.Sql/preview/2017-03-01-preview/longTermRetention.json:
```diff
@@ -0,0 +1,724 @@
+          "$ref": "#/definitions/Resource"
+        }
+      ],
+      "properties": {}
```
remove empty properties

anuchandy commented on specification/sql/resource-manager/Microsoft.Sql/preview/2017-03-01-preview/longTermRetention.json:
```diff
@@ -0,0 +1,724 @@
+          "LongTermRetentionBackups"
+        ],
+        "description": "Gets a long term retention backup.",
+        "operationId": "LongTermRetentionBackups_Get",
```
there is no `LongTermRetentionBackups_CreateOrUpdate` operation, just curious is it the case that service just creates backups at regular interval without explicitly asking for it.

trgrie replied:
Yes, Azure SQL DB takes full backups (which is what we store as part of LTR) once a week.

anuchandy commented:
Had a chat with @trgrie, service team is rolling out this feature, this needs to be merged around Jan/28 so adding "don't merge" prefix to title.

trgrie committed:
Remove empty properties and update retention description

This update is a response to the few comments that Anu left on the PR. It removes an empty properties entry and updates the retention descriptions to explain that they use an ISO 8601 standard.

anuchandy commented:
@ravbhatnagar can you please  take a look?

ravbhatnagar commented on specification/sql/resource-manager/Microsoft.Sql/preview/2017-03-01-preview/longTermRetention.json:
```diff
@@ -0,0 +1,723 @@
+    "application/json"
+  ],
+  "paths": {
+    "/subscriptions/{subscriptionId}/providers/Microsoft.Sql/locations/{locationName}/servers/{serverName}/databases/{databaseName}/longTermRetentionBackups/{backupName}": {
```
longtermRetentionbackups dont live inside a resource group? This does not seem right since the server and database are both under resource group and this is a child resource under databases.

trgrie replied:
This is because the customer can drop their database but we still need a way to get their backups for restore for up to 10 years, well beyond how long CMS will keep information about a dropped database around.

ravbhatnagar replied:
Sounds good!

ravbhatnagar commented on specification/sql/resource-manager/Microsoft.Sql/preview/2017-03-01-preview/longTermRetention.json:
```diff
@@ -0,0 +1,723 @@
+      "description": "Properties of a long term retention backup",
+      "type": "object",
+      "properties": {
+        "serverName": {
```
Instead of just the name, this should be the fully qualified ARM resource ID.

jaredmoo replied:
I'm not sure that this is possible because the server might be deleted.

ravbhatnagar replied:
got it!

ravbhatnagar commented on specification/sql/resource-manager/Microsoft.Sql/preview/2017-03-01-preview/longTermRetention.json:
```diff
@@ -0,0 +1,723 @@
+          "type": "string",
+          "readOnly": true
+        },
+        "databaseName": {
```
same as above - fully qualified arm resource id.

jaredmoo replied:
Same issue - database might be deleted

ravbhatnagar replied:
got it

ravbhatnagar commented on specification/sql/resource-manager/Microsoft.Sql/preview/2017-03-01-preview/longTermRetention.json:
```diff
@@ -0,0 +1,723 @@
+      "type": "object",
+      "properties": {
+        "weeklyRetention": {
+          "description": "The weekly retention policy for an LTR backup in an ISO 8601 format.",
```
add format- date-time for all the 3 properties below.

jaredmoo replied:
This is ISO **duration** format. e.g. "P1W" representing 1 week duration. Not actually a datetime value.

ravbhatnagar commented:
Signing off

anuchandy commented:
@trgrie just checking,  as per plan service was suppose to go-live by jan end, if plan is changed then please let us know new date. I can update title accordingly.

trgrie commented:
I don’t have an exact date yet, but I will find out and let you know when I know.

Thanks,

Trevor Grieger

From: Anu Thomas Chandy [mailto:notifications@github.com]
Sent: Thursday, February 8, 2018 3:21 PM
To: Azure/azure-rest-api-specs <azure-rest-api-specs@noreply.github.com>
Cc: Trevor Grieger <trgrie@microsoft.com>; Mention <mention@noreply.github.com>
Subject: Re: [Azure/azure-rest-api-specs] [Don't Merge, Service Rolling out ~Jan/28/18] Adding API Specs for Long Term Retention V2 (#2256)


@trgrie<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Ftrgrie&data=04%7C01%7Ctrgrie%40microsoft.com%7Cffe9cc42e51d4de70d9e08d56f4aa409%7Cee3303d7fb734b0c8589bcd847f1c277%7C1%7C0%7C636537288727634484%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwifQ%3D%3D%7C-1&sdata=35sOQ%2F9AdVv11zsytX55rIaXXaQUmKprXiILthA3d7I%3D&reserved=0> just checking, as per plan service was suppose to go-live by jan end, if plan is changed then please let us know new date. I can update title accordingly.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FAzure%2Fazure-rest-api-specs%2Fpull%2F2256%23issuecomment-364282639&data=04%7C01%7Ctrgrie%40microsoft.com%7Cffe9cc42e51d4de70d9e08d56f4aa409%7Cee3303d7fb734b0c8589bcd847f1c277%7C1%7C0%7C636537288727644488%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwifQ%3D%3D%7C-1&sdata=gDq74J1WrG4Ch%2FRLmlArYXMkPjJHPC2Svf9HwDMIXP4%3D&reserved=0>, or mute the thread<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAVIOzS0S5Sonqc1oEbKQgA4Aa6SUcjEqks5tS4FlgaJpZM4RbUOU&data=04%7C01%7Ctrgrie%40microsoft.com%7Cffe9cc42e51d4de70d9e08d56f4aa409%7Cee3303d7fb734b0c8589bcd847f1c277%7C1%7C0%7C636537288727644488%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwifQ%3D%3D%7C-1&sdata=WHcw0SwRf8RNBUHoXnKYi8DGsQ08bXmpv4%2FA4wUDZKA%3D&reserved=0>.

trgrie committed:
Description and Entity Name Changes

Made some slight changes to the entity names to better align with LTR V1 and added some better descriptors to some property descriptions.

AutorestCI commented:
# Automation for azure-sdk-for-python

AutorestCI commented:
# Automation for azure-sdk-for-go

trgrie committed:
Fixing Incorrect Description

 The weekOfYear parameter is not in an ISO 8601 format, it is just an integer between 1 and 52. Fixing that accordingly.

trgrie committed:
Remove Conflicting Definitions

Need to remove backupLongTermRetentionPolicies so it doesn't conflict with the new backupLongTermRetentionPolicies.

trgrie committed:
Remove Reference to backupLongTermRetentionPolicies.json

Need to remove the reference to backupLongTermRetentionPolicies.json in the readme.

trgrie committed:
Update ResourceId

Updated the resource ID for longTermRetentionBackups so servers and databases are now longTermRetentionServers and longTermRetentionDatabases, respectively. This better aligns with REST.

trgrie committed:
Re-Adding LongTermRetentionPolicyGet.json

LongTermRetentionPolicyGet.json was accidentally deleted in the last commit.

anuchandy committed:
Merge branch 'master' into master

AutorestCI commented:
# Automation for azure-libraries-for-java

trgrie committed:
Fix Travis CI Errors

Made some changes to the examples to fix the Travis CI errors.

trgrie committed:
Pulling Jared's readme.md changes and updating LongTermRetentionPolicies to BackupLongTermRetentionPolicies

AutorestCI commented:
# Automation for azure-sdk-for-node

trgrie committed:
Updating more references to BackupLongTermRetentionPolic(y|ies)

trgrie committed:
Re-adding backupLongTermRetentionPolicies.json

jaredmoo commented on .travis.yml:
```diff
@@ -5,7 +5,6 @@ services:
   - docker
 env:
   matrix:
-    - MODE=branchStrategy
```
undo this

trgrie committed:
removing .travis.yml that was accidentally added

trgrie committed:
Revert "removing .travis.yml that was accidentally added"

This reverts commit 3b80fa0ba34e180a5d215f3a9bd8d884991ec639.

trgrie committed:
Revert "Re-adding backupLongTermRetentionPolicies.json"

This reverts commit b98cf6c5070d9fa6ebb2aa173462d3239d696694.

trgrie committed:
undoing accidental .travis.yml

trgrie committed:
Merge branch 'master' into master

trgrie committed:
Fixing errors in model PR_ONLY=true Travis CI run

trgrie committed:
Merge branch 'master' of https://github.com/trgrie/azure-rest-api-specs

trgrie committed:
Fixing another Travis CI error

trgrie committed:
Removing .travis.yml from the change and changing LongTermRetentionPolicy to BackupLongTermRetentionPolicy to keep it consistent

trgrie committed:
resetting backupLongTermRetentionPolicies.json

trgrie committed:
Update more references of LongTermRetentionPolicy to BackupLongTermRetentionPolicy

trgrie committed:
updating readme.md comment
